File: code/frontend/src/services/review-service.ts
import api from '@/lib/axios';
import { PageResponse } from '@/types/api';
import { ReviewRequest, ReviewResponse, AdminReviewAnalyticsResponse } from '@/types/backend';

export const reviewService = {
  // Public API
  getProductReviews: async (productId: number, page: number = 0, size: number = 10, sort: string = 'newest'): Promise<PageResponse<ReviewResponse>> => {
    const res = await api.get(`/products/${productId}/reviews`, {
      params: { page, size, sort }
    });
    return res.data;
  },

  getReviewDetail: async (id: number): Promise<ReviewResponse> => {
    const res = await api.get(`/reviews/${id}`);
    return res.data;
  },

  // Renter API
  createReview: async (data: ReviewRequest): Promise<ReviewResponse> => {
    const res = await api.post('/reviews', data).catch(err => {
      console.error("API Error Response:", err.response?.data);
      throw err;
    });
    return res.data;
  },

  // Admin API
  getAllReviewsAdmin: async (page: number = 0, size: number = 10): Promise<PageResponse<ReviewResponse>> => {
    const res = await api.get('/admin/reviews', {
      params: { page, size }
    });
    return res.data;
  },

  deleteReviewAdmin: async (id: number): Promise<void> => {
    await api.delete(`/admin/reviews/${id}`);
  },

  getReviewAnalyticsAdmin: async (): Promise<AdminReviewAnalyticsResponse> => {
    const res = await api.get('/admin/reviews/analytics');
    return res.data;
  }
};

File: code/backend/src/main/java/com/ioc/internship/service/ReviewService.java
package com.ioc.internship.service;

import com.ioc.internship.dto.request.ReviewRequest;
import com.ioc.internship.dto.response.AdminReviewAnalyticsResponse;
import com.ioc.internship.dto.response.ReviewResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ReviewService {
    ReviewResponse createReview(String email, ReviewRequest request);
    
    Page<ReviewResponse> getProductReviews(Long productId, Pageable pageable, String sort);
    
    ReviewResponse getReviewDetail(Long id);
    
    Page<ReviewResponse> getAllReviewsAdmin(Pageable pageable);
    
    void deleteReviewAdmin(Long id);
    
    AdminReviewAnalyticsResponse getReviewAnalytics();
    
    void hideReview(Long id, String reason, Long adminId);
    
    void restoreReview(Long id);
}
